text/0513-app-specific-staging.md

```diff
@@ -0,0 +1,281 @@
+    * Docker caching relies on pulling the “latest” image from a repository and skipping layers that were already built.
+      This doesn’t work if images built off of various different Dockerfiles are in the same repository.
+    * Lifecycle policies cannot be used because different images from potentially different applications with very
+      different life cycles are all in the same repository. The same was already true for S3, but the problem is
+      less severe because S3 is pretty cheap while ECR is not.
+    * Some people were using the V1 Docker image publishing mechanism not as a vehicle for uploading Docker images to be used
+      by the CDK’s CloudFormation deployment, but simply as a mechanism for building and publishing Docker images, to be
+      used by a completely different deployment later. The lack of control over the target ECR repository breaks this
+      use case (required the development of an `aws-ecr-deployments` construct module, which does give the necessary
+      control but racks up costs by doubling ECR storage requirements, and still does not allow staging resource cleanup).
+    * We always create an empty ECR repository because we cannot know whether apps deployed into the account will need
+      it or not, so the ECR repository may go unused. AWS Security Hub will throw warnings about empty ECR repositories,
+      which makes customers uneasy.
+* Bootstrap stacks are expected to be account-wide, and mix assets from all applications. Some customers that deploy
+  multiple applications into the same account are very sensitive to this mixing, and would rather keep these resources
+  separate. They can do multiple bootstrap stacks in the same account, but this is all a bit onerous.
+
+## A new proposal: application specific staging resources
+
+The bootstrap stack contains two classes of resources: staging resources, which hold assets (bucket and ECR repo), and
+roles, which allow for unattended (CI/CD) and cross-account access. In the new proposal, we will separate out the
+staging resources from the roles. Roles will still be bootstrapped (if used), but staging resources will not.
+
+* Staging resources will be created on a per-CDK app basis. We will create one S3 bucket with different object prefixes
+  for different types of assets (see Appendix A: two types of assets), and an ECR repository per Docker image. Resource
+  access roles can also be created on an as-needed basis. This solves the problems of asset resources of different
+  applications mixing together, and it would also remove the need for garbage collection by allowing use of life cycle
+  rules.
+* Since the roles are now the only things that need to be bootstrapped, that will have a number of advantages:
+    * Bootstrapping will be faster since the heavy resource of a KMS key is no longer involved.
+    * Because roles are a global resource, every account now only needs to be bootstrapped once. First of all the lack
```

Comment:
Can you add a section to this RFC that goes over the changes needed to the bootstrap template? How will users opt-in to the new bootstrap template? Is there a migration path?